### PR TITLE
Improve graph execution stats reporting

### PR DESCRIPTION
**Pull request by RyanJDick**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [ ] Feature
- [ ] Bug Fix
- [x] Optimization
- [ ] Documentation Update
- [ ] Community Node Submission

## Have you discussed this change with the InvokeAI team?
- [x] Yes
- [ ] No, because:

      
## Have you updated all relevant documentation?
- [x] Yes: N/A
- [ ] No

## Description

This PR makes the following improvements to the graph execution stats reporting:
- Report graph execution _wall time_ in addition to node execution time.
- Fix a bug in the reporting of the RAM diff from this line: `RAM used by InvokeAI process: 4.10G (+1.439G)`. Previously, the diff only reflected the diff during the last node of the graph. Now the diff reflects the entire graph execution.
- Improve general code readability by removing the `StatsContext` layer.
- Reduce the number of unnecessary calls to `graph_execution_manager.get(graph_id)`. This is called many times during graph execution and takes a significant amount of time. It was previously being called in the stats service for every node invocation. It is now only called on rare occasions from the stats service (during pruning of stale stats). There are still a number of other places making excessive calls to `graph_execution_manager.get(graph_id)`, but this is a step in the right direction at least. 

### Stats log before change:

```bash
[2024-01-11 15:27:06,090]::[InvokeAI]::INFO --> Graph stats: d29ccfd6-ebba-4f93-905f-4acb80a45f70
[2024-01-11 15:27:06,090]::[InvokeAI]::INFO -->                           Node   Calls  Seconds  VRAM Used
[2024-01-11 15:27:06,090]::[InvokeAI]::INFO -->              main_model_loader     1     0.007s     0.000G
[2024-01-11 15:27:06,090]::[InvokeAI]::INFO -->                      clip_skip     1     0.009s     0.000G
[2024-01-11 15:27:06,091]::[InvokeAI]::INFO -->                         compel     2     1.212s     0.246G
[2024-01-11 15:27:06,091]::[InvokeAI]::INFO -->                          noise     1     0.007s     0.244G
[2024-01-11 15:27:06,091]::[InvokeAI]::INFO -->                denoise_latents     1     3.661s     2.040G
[2024-01-11 15:27:06,091]::[InvokeAI]::INFO -->                  core_metadata     1     0.006s     1.858G
[2024-01-11 15:27:06,091]::[InvokeAI]::INFO -->                            l2i     1     0.495s     3.156G
[2024-01-11 15:27:06,091]::[InvokeAI]::INFO -->               linear_ui_output     1     0.010s     2.152G
[2024-01-11 15:27:06,091]::[InvokeAI]::INFO --> TOTAL GRAPH EXECUTION TIME:    5.406s
[2024-01-11 15:27:06,091]::[InvokeAI]::INFO --> RAM used by InvokeAI process: 4.05G (+0.000G)
[2024-01-11 15:27:06,091]::[InvokeAI]::INFO --> RAM used to load models: 1.99G
[2024-01-11 15:27:06,092]::[InvokeAI]::INFO --> VRAM in use: 2.152G
[2024-01-11 15:27:06,092]::[InvokeAI]::INFO --> RAM cache statistics:
[2024-01-11 15:27:06,092]::[InvokeAI]::INFO -->    Model cache hits: 2
[2024-01-11 15:27:06,092]::[InvokeAI]::INFO -->    Model cache misses: 5
[2024-01-11 15:27:06,092]::[InvokeAI]::INFO -->    Models cached: 5
[2024-01-11 15:27:06,092]::[InvokeAI]::INFO -->    Models cleared from cache: 0
[2024-01-11 15:27:06,092]::[InvokeAI]::INFO -->    Cache high water mark: 1.99/8.00G
```

### Stats log after change:

```bash
[2024-01-11 15:31:07,088]::[InvokeAI]::INFO --> Graph stats: 9b8f974e-e65c-451e-be10-6574c4fe6251
                          Node   Calls  Seconds  VRAM Used
             main_model_loader     1     0.008s     0.000G
                     clip_skip     1     0.013s     0.000G
                        compel     2     1.225s     0.246G
                         noise     1     0.009s     0.244G
               denoise_latents     1     3.692s     2.040G
                 core_metadata     1     0.006s     1.858G
                           l2i     1     0.465s     3.156G
              linear_ui_output     1     0.010s     2.152G
TOTAL GRAPH EXECUTION TIME:   5.427s
TOTAL GRAPH WALL TIME:   5.591s
RAM used by InvokeAI process: 4.08G (+1.947G)
RAM used to load models: 1.99G
VRAM in use: 2.152G
RAM cache statistics:
   Model cache hits: 2
   Model cache misses: 5
   Models cached: 5
   Models cleared from cache: 0
   Cache high water mark: 1.99/8.00G
```

Things to note:
- `TOTAL GRAPH WALL TIME` is reported. It is 0.164s longer than the `TOTAL GRAPH EXECUTION TIME` in this case.
- The diff in `RAM used by InvokeAI process: 4.08G (+1.947G)` represents the entire graph.

## Performance Tests

I ran a simple scalar addition workflow with an iterate node for 100 iterations. Before this change it took ~13secs of wall time, after the change it took ~11secs of wall time. Not as big of a gain as I had hoped. There's still lots of overhead to adress in the graph engine.

## Added/updated tests?

- [ ] Yes
- [x] No

### CODE REVIEW

File: invokeai/app/services/invocation_processor/invocation_processor_default.py [hunk 1]
```diff
@@ -195,6 +194,7 @@ def __process(self, stop_event: Event):
                             error=traceback.format_exc(),
                         )
                 elif is_complete:
+                    self.__invoker.services.performance_statistics.log_stats(graph_execution_state.id)
```
Comment (RyanJDick) [inline invokeai/app/services/invocation_processor/invocation_processor_default.py:197]: I'm not super familiar with this code. @psychedelicious is there a better place for this?

Comment (psychedelicious) [reply invokeai/app/services/invocation_processor/invocation_processor_default.py:197]: This is earliest we know a graph is complete, I think it's a reasonable spot. The only other area would be in `invokeai/app/services/session_queue/session_queue_sqlite.py` in the handlers for complete/failed/canceled session queue items.

Unrelated, but are we resetting the stats correctly when a graph is canceled?

**Comment by psychedelicious** [comment]
> The diff in RAM used by InvokeAI process: 4.08G (+1.947G) represents the entire graph.

How does the diff represent the entire graph? Like, we were at ~2GB baseline and another ~2GB was used during execution?

File: invokeai/app/services/invocation_processor/invocation_processor_default.py (hunk 1 above)
Comment (RyanJDick) [reply invokeai/app/services/invocation_processor/invocation_processor_default.py:197]: We try to reset when we know that we should, but I don't have high confidence that we reset in all cancellation and error cases. This is why I've kept `_prune_stale_stats(...)` for now. Maybe its just my lack of familiarity with `invocation_processor_default.py`, but I had a hard time tracing all of the possible termination paths.

**Comment by RyanJDick** [comment]
> > The diff in RAM used by InvokeAI process: 4.08G (+1.947G) represents the entire graph.
> 
> How does the diff represent the entire graph? Like, we were at ~2GB baseline and another ~2GB was used during execution?

`ram_change = last_node.end_ram_gb - first_node.start_ram_gb`

It would be more informative if we could report peak RAM, but I'm not aware of a way to record that without impacting performance. This at least seemed better than just reporting `last_node.end_ram_gb - last_node.start_ram_gb`.

File: invokeai/app/services/invocation_processor/invocation_processor_default.py (hunk 1 above)
Comment (psychedelicious) [reply invokeai/app/services/invocation_processor/invocation_processor_default.py:197]: It's not lack of familiarity - it's a bit convoluted. Needs a bit of TLC.

**Review by psychedelicious — APPROVED** [review]

**Pull request merged**